typescript/solver/solvers/onChain/filler.ts

```diff
@@ -0,0 +1,134 @@
+import { Wallet } from "@ethersproject/wallet";
+import { chainMetadata } from "@hyperlane-xyz/registry";
+import { MultiProvider } from "@hyperlane-xyz/sdk";
+import { bytes32ToAddress, ensure0x } from "@hyperlane-xyz/utils";
+
+import { MNEMONIC, PRIVATE_KEY } from "../../config.js";
+import { DestinationSettler__factory } from "../../contracts/typechain/factories/DestinationSettler__factory.js";
+import { Erc20__factory } from "../../contracts/typechain/factories/Erc20__factory.js";
+import { logDebug, logGreen } from "../../logger.js";
+import type { OpenEventArgs, ResolvedCrossChainOrder } from "../../types.js";
+import { getChainIdsWithEnoughTokens, settleOrder } from "./utils.js";
+
+export const create = () => {
+  const { multiProvider } = setup();
+
+  return async function onChain({ orderId, resolvedOrder }: OpenEventArgs) {
+    logGreen("Received Order:", orderId);
+
+    const { fillInstructions, maxSpent } = await selectOutputs(
+      resolvedOrder,
+      multiProvider,
+    );
+
+    await fill(orderId, fillInstructions, maxSpent, multiProvider);
+
+    logGreen(`Filled ${fillInstructions.length} leg(s) for:`, orderId);
+  };
+};
+
+function setup() {
+  if (!PRIVATE_KEY && !MNEMONIC) {
+    throw new Error("Either a private key or mnemonic must be provided");
+  }
+
+  const multiProvider = new MultiProvider(chainMetadata);
+  const wallet = PRIVATE_KEY
+    ? new Wallet(ensure0x(PRIVATE_KEY))
+    : Wallet.fromMnemonic(MNEMONIC!);
+  multiProvider.setSharedSigner(wallet);
+
+  return { multiProvider };
+}
+
+// We're assuming the filler will pay out of their own stock, but in reality they may have to
+// produce the funds before executing each leg.
+async function selectOutputs(
```

Comment:
Would probably name this something more like "canFill" or "prepareIntent" or something to not make it 7683 specific

Reply:
This reminds me that we will need some kind of `Result` ish type probably throughout. I always go back and forth on doing this in a language that doesn't support this natively, but it just seems so useful. We use zod for validation (which btw i would be in favor of adopting here too), but a simple `{success: bool, result: T, error: TError}` seems maybe like a good compromise

Reply:
Sure thing, we'll do it before merge